Fermichem part 2.1: Adds a reaction/reagent lookup GUI!

Thalpy opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Adds a new GUI that can be opened from the chem dispenser that will let you look up all reactions and reagents presently in the game (except secret recipes).

For the colours:
purple - clicking this will give you a recipe
blue - clicking this will give you a reagent
green - clicking this will turn it off (or it's also green if the reagent is present in the associated beaker the UI is linked to)
red - clicking this will turn it on 

I tend to get verbose so here's a video instead!

https://user-images.githubusercontent.com/33956696/107724825-aca48480-6cdc-11eb-9ba3-800627fc937e.mp4

And a picture:
![image](https://user-images.githubusercontent.com/33956696/107725010-1d4ba100-6cdd-11eb-84ee-3cebb9754f37.png)

## Why It's Good For The Game

Fermichem changed a lot, and I want to make it possible for people to see reaction conditions easily rather than having to alt tab to a wiki constantly. I don't think this will replace the wiki though - this is more a condensed version for finding reaction conditions, whereas the wiki has more information.

## Changelog
:cl:
add: Adds a new reaction lookup window to the chem dispenser
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Thalpy committed:
reverts changelog

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Update code/modules/mob/living/carbon/life.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Thalpy committed:
Avoid type checking reagent_list

Co-authored-by: Rohesie <rohesie@gmail.com>

Thalpy committed:
Big batch of requested changes

Thalpy committed:
Merge branch 'fermichemp1' of https://github.com/Thalpy/tgstation into fermichemp1

Thalpy committed:
Fixing compile errors/bits I missed

Thalpy committed:
Minor logic flaw

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into fermichemp1

Thalpy committed:
camels and yamls

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into fermichemp1

Thalpy committed:
Code quality and updates tgui and life to liver

Thalpy committed:
Cleaning up comments

Thalpy committed:
Few more things I noticed and antropod pointed out

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Maybe this will improve pH feel? Very undecided

Thalpy committed:
Converts drinks

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Few fixes and tweaks (see above)

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into fermichemp1

Thalpy committed:
Sawa  few things that should be clearer

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Apply part 1 of suggestions from code review

The autoresolve ones only

Co-authored-by: Rohesie <rohesie@gmail.com>

Thalpy committed:
Update code/modules/reagents/chemistry/items.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Thalpy committed:
Review changes p2

Thalpy committed:
Gotta test things now

Thalpy committed:
Guess who forgot to compile test first

Thalpy committed:
Sets testing enviroment for equilibrium.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Thalpy committed:
Apply part 1 of suggestions from code review Mk.3

Co-authored-by: Rohesie <rohesie@gmail.com>

Thalpy committed:
I messed up with stash

Thalpy committed:
Ri to inverse_reagent

Co-authored-by: Rohesie <rohesie@gmail.com>

Thalpy committed:
More fixes

Thalpy committed:
Merge branch 'fermichemp1' of https://github.com/Thalpy/tgstation into fermichemp1

Thalpy committed:
Compile check

Thalpy committed:
LAZY UPDATE LAZY OPERATIONS

Thalpy committed:
Linter

Thalpy committed:
Lets null the reference too

Thalpy committed:
LIFO

Thalpy committed:
Should be all fixed now

Thalpy committed:
Revisions, fixes and locker additions

Thalpy committed:
I hit ctrl + s without the ctrl

Thalpy committed:
Fixes the ability to eat endless pHpaper slips with no consequence

Thalpy committed:
Requested list changes

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Null list and missing catalsyst

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge remote-tracking branch 'upstream/master' into fermichemp1

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into fermichemp1

Thalpy committed:
Fixes icon conflict

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Mid tweak - need to atomise fix

Thalpy committed:
The great cheese fix

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into fermichemp1

Thalpy committed:
Triple checking that food can't and won't crash + review comments

Thalpy committed:
Fixes duplicate gravy and minorly adjusts randomized chem

Thalpy committed:
Formatting special reaction paper

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Lets not ruin this for anyone who gets it

Thalpy committed:
Final pass

Thalpy committed:
Forgot about cold overheats and reduced text spam

Thalpy committed:
The hercuri condition

Thalpy committed:
Okay, we'll see if people can figure this out

Thalpy committed:
oops commented out ifdef

Thalpy committed:
Removes forgotten debug_worlds and var/mob

Thalpy committed:
Fixes message issues

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Requested changes + heirlooms (before pH rename)

Thalpy committed:
Monster pH to ph change

Thalpy committed:
Hopefully this doesn't break everything

Thalpy committed:
Potential super hotfix for smoke problem

Thalpy committed:
I cannot believe I forgot REACTION_INSTANT

Thalpy committed:
oops

Thalpy committed:
One more

Thalpy committed:
How on earth did I forget these flags aaa

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Just in case --fix breaks everything

Thalpy committed:
Fix worked, wow

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Okay bed, bug fixing tomorrow

Thalpy committed:
Apply suggestions from code review Mk.4

Co-authored-by: Rohesie <rohesie@gmail.com>
Co-authored-by: Mothblocks <35135081+Jared-Fogle@users.noreply.github.com>

Thalpy committed:
Part 2 of Mk.4

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Attempts to fix linter problems - and I made these reactions far too lethal and also attemped to create a subsystem

Thalpy committed:
Wow I can't believe it works first try and it was that easy, and it even improves performance by a factor of 2!!

Thalpy committed:
Update _compile_options.dm

Thalpy committed:
Okay, tommorrow more bugfixes, reaction tweaks and a lil testing. Then it should be good for a PR update!

Thalpy committed:
Chem heater UI sync code

Thalpy committed:
Update _compile_options.dm

Thalpy committed:
Magically works(???)

Thalpy committed:
Few more tweaks and fixes

Thalpy committed:
Making sure meth goes boom

Thalpy committed:
Fixes clamp

Thalpy committed:
UI fix and I don't need the flag check anymore

Thalpy committed:
I changed my mind about it, I kind of like making no react beakers useful

Thalpy committed:
What is tgstation.mdme.dme ? (also does the change I just said above)

Thalpy committed:
Oops I did it backwards

Thalpy committed:
Meth crash is fixed

Thalpy committed:
Fixes buffer error

Thalpy committed:
Smol fixes

Thalpy committed:
Apply suggestions from code review Mk5

Co-authored-by: Rohesie <rohesie@gmail.com>

Thalpy committed:
Requested by hand changes

Thalpy committed:
Merge branch 'fermichemp1' of https://github.com/Thalpy/tgstation into fermichemp1

Thalpy committed:
Fixing compile checks

Thalpy committed:
Oops I missed one

Thalpy committed:
Adds UI checks

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Edits from review

Thalpy committed:
Forgot signal handlier

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Requested changes

Thalpy committed:
Indents

Thalpy committed:
Linter

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into fermichemp1

Thalpy committed:
Delete slnx.sqlite

Thalpy committed:
Double procs

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into fermichemp1

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Adjusts for world.time slowdown difference between testing and live - speeds things up more.

Thalpy committed:
Update chem_heater.dm

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into fermichemp1

Thalpy committed:
Moving dm to more appropriate folder

Thalpy committed:
Fixes runtimes observed in logs

Thalpy committed:
Speeds up drinks too

Thalpy committed:
This shouldn't be needed - realised 0 will equal 0

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
First steps

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
All code required written

Thalpy committed:
Bed - almost works

Thalpy committed:
Oops broke everything

Thalpy committed:
Master of functional ui disaster

Thalpy committed:
Feature complete??

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Works! Needs a bit more testing and tweaking but it's functional

Thalpy committed:
Smol fixes brfore hopping branch

Thalpy committed:
Merge branch 'fermichemp1'

Thalpy committed:
Revert "Merge branch 'fermichemp1'"

This reverts commit 3791210b3aacb4ad51c013802a3ac0909ea9ee4a, reversing
changes made to e6f2b07a69284129e6a2188aa97b01adc2a97c02.

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
slimes and list consensing

Thalpy committed:
Mid tweak

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Mid tagging

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Finished tags

Thalpy committed:
Ready - except linters

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into reaction-lookup-attempt

Thalpy committed:
Cleaning up for NO_OVERHEAT

Thalpy committed:
Tuning it before I put up the PR

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Oops !=

Thalpy commented:
I'm having a few troubles with .js linters - I've 1 too many consts, and one of my lines is 1 char over. I figure I'll likely have to change bits about it anyways because it's a define dump to deal with bitflags, so hopefully they'll be cleaned up from feedback.

```
     7:25  error  Arrow function has a complexity of 51. Maximum allowed is 50  complexity
  522:1   error  This line has a length of 81. Maximum allowed is 80           max-len
```

Thalpy committed:
Adds a wikibutton

Thalpy committed:
Fixes strange graphs

Thalpy committed:
removes a bunch of unneeded stuff in ui code

Thalpy committed:
cleaning up the code - except I broke it

Hamcha commented:
As someone who spent a lot of time making a better wiki client just because of how vital having a reference is, I cannot thank you enough for this!

However, I feel like for more complex reagents it might be just as finnicky to lookup parent reagents as vanilla wiki is.
In my script, I auto-nested all recipes like this:

<img src="https://i.imgur.com/4z6SmQd.png">

It is not pretty but in my experience it's way too handy.

Thalpy commented:
> 
> 
> As someone who spent a lot of time making a better wiki client just because of how vital having a reference is, I cannot thank you enough for this!
> 
> However, I feel like for more complex recipes it might be just as finnicky to lookup parent reagents as vanilla wiki is.
> In my script, I auto-nested all recipes like this:
> 
> ![](https://camo.githubusercontent.com/6a8085ae6570939567fc9d26b4710760bbd535aef08eeb3e5f001341babc7147/68747470733a2f2f692e696d6775722e636f6d2f347a36536d51642e706e67)
> 
> It is not pretty but in my experience it's way too handy.

I wanted to do something similar to that, but I couldn't figure out how to have miltiple tooltips appear to the right of a tooltip (i.e. nesting tooltips). Do you have any idea how to do that? Presently it shows one level down as a tooltip hover.

Hamcha commented:
Multiple tooltips not working is why I moved away from tooltips. (In fact, most of the work I did on the script was just making the tooltips *not* tooltips and inline them to the content of the page).

I have near zero experience with tgui, so no idea how to help you there, but from a design perspective: perhaps some sort of bookmark system might help alleviate the issue?

Jared-Fogle commented on tgui/packages/tgui/interfaces/Reagents.js:
```diff
@@ -1,9 +1,71 @@
 import { Window } from '../layouts';
 import { map } from 'common/collections';
 
+const BRUTE = 1 << 0;
+const BURN = 1 << 1;
+const TOXIN = 1 << 2;
+const OXY = 1 << 3;
+const CLONE = 1 << 4;
+const HEALING = 1 << 5;
+const DAMAGING = 1 << 6;
+const EXPLOSIVE = 1 << 7;
+const OTHER = 1 << 8;
+const DANGEROUS = 1 << 9;
+const EASY = 1 << 10;
+const MODERATE = 1 << 11;
+const HARD = 1 << 12;
+const ORGAN = 1 << 13;
+const DRINK = 1 << 14;
+const FOOD = 1 << 15;
+const SLIME = 1 << 16;
+const DRUG = 1 << 17;
+const UNIQUE = 1 << 18;
+const CHEMICAL = 1 << 19;
+const PLANT = 1 << 20;
+const COMPETITIVE = 1 << 21;
```
These should be passed down through `ui_static_data` so we don't have two places to update if we want to change the flags.

stylemistake replied:
true

Thalpy replied:
Hmm, do I still set them up as consts then?

stylemistake replied:
return them via ui_static_data

Thalpy committed:
Fixed it somehow!

Thalpy committed:
Okay, fixed things

Thalpy commented:
> 
> 
> Multiple tooltips not working is why I moved away from tooltips. (In fact, most of the work I did on the script was just making the tooltips _not_ tooltips and inline them to the content of the page).
> 
> I have near zero experience with tgui, so no idea how to help you there, but from a design perspective: perhaps some sort of bookmark system might help alleviate the issue? If you could jump back to your main recipe while seeing the nested reactions, not having a tree view probably becomes a non-issue (and it would work better with beaker sync imo)

I like the idea of bookmarks, I could just get a big list of all component reagents, but with how reactions work now it would likely not work out if you blindly followed it (since everything will react if it gets the chance). I'll try to put that in when I've reworked the js code from maintainer feedback.

Thalpy committed:
No free loot today

Thalpy committed:
Revert "No free loot today"

This reverts commit 05701c0f107f97d10bbb223ba858a93e0e30694e.

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

actions-user committed:
Automatic changelog compile [ci skip]

Tlaltecuhtli commented:
can the recipe you want to do be like uhh shown on the window with main chem dispenser window once clicked? like you click ephedrine and the main window shows 1 oil 1 dieth 1 sugar 1 hydrogen as suggestion which you can click X to remove it once ur done

Thalpy committed:
I broke bookmarks

Thalpy committed:
Basic bookmarking

Thalpy committed:
Linter wants strange things sometimes

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into reaction-lookup-attempt

Thalpy commented:
Okay, I added a bookmarks function. It saves itself between instances and beakers too!

https://user-images.githubusercontent.com/33956696/107855816-88e85800-6e1c-11eb-8705-29c00df996e4.mp4

> 
> 
> can the recipe you want to do be like uhh shown on the window with main chem dispenser window once clicked? like you click ephedrine and the main window shows 1 oil 1 dieth 1 sugar 1 hydrogen as suggestion which you can click X to remove it once ur done

Hmm, I'm not sure how to fit it into the dispenser ui cleanly, but I have a similar idea, I'll just get maintainer approval first.

Thalpy committed:
Minor cleaning up

Thalpy committed:
Cleaning up and catching a bug

Thalpy committed:
Linter please stop bullying me

Thalpy committed:
Linter you scoundrel

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into reaction-lookup-attempt

Thalpy committed:
updates comments to be more accurate and highlights recipes in chem dispenser

Thalpy committed:
cleans up ui code

Thalpy committed:
Fixes highlighting bug

Thalpy committed:
Tweaks to make it look nicer

Thalpy committed:
Linters my old nemesis

Thalpy commented:
Okay so instead of having recipes appear on the chem dispenser, I've instead made it so that if the dispenser has a reagent in the selected recipe it will highlight it.

stylemistake commented on tgui/packages/tgui/interfaces/common/ReagentLookup.js:
```diff
@@ -0,0 +1,123 @@
+            })} />
+        </LabeledList.Item>
+      </LabeledList>
+    ) || (
+      <Box>
+        No reagent selected!
+      </Box>
+    )
```
Use early return in this component.

```
if (!reagents) {
  return (
    your box
  );
}

return (
  rest of UI
);
```

Thalpy replied:
Done so

stylemistake commented on tgui/packages/tgui/interfaces/common/ReagentLookup.js:
```diff
@@ -0,0 +1,123 @@
+            color="teal"
+            tooltip="Open the associated wikipage for this reagent."
+            tooltipPosition="left"
+            onClick={() => 
+              Byond.command(`wiki Guide_to_chemistry#${reagent.name}`)} />
```
```suggestion
            onClick={() => {
              Byond.command(`wiki Guide_to_chemistry#${reagent.name}`);
            }} />
```

stylemistake commented on tgui/packages/tgui/interfaces/common/ReagentLookup.js:
```diff
@@ -0,0 +1,123 @@
+                <Button
+                  key={reagent.inverseReagent}
+                  icon="vial"
+                  content={reagent
+                    .inverseReagent}
```
```suggestion
                  content={reagent.inverseReagent}
```

stylemistake commented on tgui/packages/tgui/interfaces/common/RecipeLookup.js:
```diff
@@ -0,0 +1,210 @@
+        </LabeledList.Item>
+      </LabeledList>
+    ) || (
+      <Box>
+        No reaction selected!
+      </Box>
```
Same can be done here, good candidate for early return.

Thalpy replied:
same as above

stylemistake commented on tgui/packages/tgui/interfaces/common/RecipeLookup.js:
```diff
@@ -0,0 +1,210 @@
+              color="transparent"
+              textColor={recipe.isColdRecipe ? "red" : "white"}
+              content={recipe.isColdRecipe ? recipe.explodeTemp+"K" : recipe.tempMin+"K"}
+              tooltip={recipe.isColdRecipe ? "The temperature at which it is underheated, causing negative effects on the reaction." : "The minimum temperature needed for this reaction to start. Heating it up past this point will increase the reaction rate."}
```
Break up into multiple lines. Linter ignores these lines because they have strings, but it doesn't exempt you from at least putting newlines before `?` and `:`.

Thalpy replied:
Didn't realise I could do that! Thanks.

stylemistake commented on tgui/packages/tgui/interfaces/common/RecipeLookup.js:
```diff
@@ -0,0 +1,210 @@
+                  color="transparent"
+                  textColor={recipe.isColdRecipe ? "white" : "red"}
+                  content={recipe.isColdRecipe ? recipe.tempMin+"K" : recipe.explodeTemp+"K"}
+                  tooltip={recipe.isColdRecipe ? "The minimum temperature needed for this reaction to start. Heating it up past this point will increase the reaction rate." : "The temperature at which it is overheated, causing negative effects on the reaction."}
```
Same here

Thalpy replied:
same as above

stylemistake commented on tgui/packages/tgui/interfaces/common/RecipeLookup.js:
```diff
@@ -0,0 +1,210 @@
+          <Button
+            key={"increment_index"}
+            icon="arrow-right"
+            disabled={recipe.subReactIndex
+                            === recipe.subReactLen?true:false}
```
Are you serious?

`? true : false` moment.

```suggestion
            disabled={recipe.subReactIndex === recipe.subReactLen}
```

stylemistake commented on tgui/packages/tgui/interfaces/common/RecipeLookup.js:
```diff
@@ -0,0 +1,210 @@
+            key={"reduce_index"}
+            icon="arrow-left"
+            ml={3}
+            disabled={recipe.subReactIndex
+                            === 1 ? true : false}
```
```suggestion
            disabled={recipe.subReactIndex === 1}
```

stylemistake commented on tgui/packages/tgui/interfaces/common/RecipeLookup.js:
```diff
@@ -0,0 +1,210 @@
+          <Icon name="circle" mr={1} color={recipe.reagentCol} />
+          {recipe.name}
+          <Button
+            key={"reduce_index"}
```
Pointless key
```suggestion
```

stylemistake commented on tgui/packages/tgui/interfaces/ChemDispenser.js:
```diff
@@ -137,7 +146,7 @@ export const ChemDispenser = (props, context) => {
                 content={chemical.title}
                 tooltip={"pH: " + chemical.pH}
-                backgroundColor={hasCol ? chemical.pHCol : "primary"}
+                backgroundColor={recipeReagents.includes(chemical.id) ? "black" : hasCol ? chemical.pHCol : "default"}
```
break up this long line on `?` and `:`

Thalpy replied:
Done

Thalpy committed:
Apply suggestions from code review, thanks Style!

Co-authored-by: Aleksej Komarov <stylemistake@gmail.com>

Thalpy committed:
Updating other points

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into reaction-lookup-attempt

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Mothblocks commented on code/__DEFINES/reagents.dm:
```diff
@@ -12,12 +12,12 @@
 #define OPENCONTAINER (REFILLABLE | DRAINABLE | TRANSPARENT)
 
-// Reagent exposure methods.
+// Reagent exposure methods.s
```
s

Thalpy replied:
ctrl s without the ctrl

Thalpy committed:
Merge branch 'master' into reaction-lookup-attempt

Thalpy committed:
No s

actions-user committed:
Automatic changelog compile [ci skip]

Thalpy committed:
Adds a bookmark button to the recipe viewer

Thalpy committed:
linters and candy

Thalpy committed:
made purity values %

actions-user committed:
Automatic changelog compile [ci skip]

stylemistake commented on tgui/packages/tgui/interfaces/ChemDispenser.js:
```diff
@@ -35,14 +36,22 @@ export const ChemDispenser = (props, context) => {
           title="Status"
           buttons={(
-            <Fragment>          
+            <Fragment>
```
Replace with an empty tag, <>, </>

Thalpy replied:
Done

stylemistake commented on tgui/packages/tgui/interfaces/Reagents.js:
```diff
@@ -0,0 +1,418 @@
+    return false;
+  }
+  if (currentReagents.includes(reagent)) {
+    return true;
```
you can replace all of this function with currentReagents?.includes(reagent)

Thalpy replied:
done

stylemistake commented on tgui/packages/tgui/interfaces/Reagents.js:
```diff
@@ -0,0 +1,418 @@
+  };
+
+  const visibleReactions = master_reaction_list.filter(reaction => (
+    (selectedBitflags?matchBitflag(selectedBitflags, reaction.bitflags):true)
```
Spaces

Thalpy replied:
And done!

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' of https://github.com/Thalpy/tgstation

Thalpy committed:
Merge branch 'master' into reaction-lookup-attempt

Thalpy committed:
requested changes and fixes to addiction

Thalpy committed:
Forgot about linters

Thalpy committed:
Quick fix

Thalpy committed:
made purities not appear if 0

stylemistake committed:
Update RecipeLookup.js

stylemistake committed:
Update RecipeLookup.js

stylemistake committed:
Update ChemDispenser.js

actions-user committed:
Automatic changelog compile [ci skip]

stylemistake committed:
Update Reagents.js

I really had to dig through this one

stylemistake committed:
Update ReagentLookup.js

stylemistake committed:
Update Reagents.js

stylemistake commented:
Hey, master of copypasta, please check if this still works as intended.

Review my changes and maybe you'll learn something.

What I don't like about your code in general:
- You use Button as a glorified holder for a tooltip that otherwise does nothing. You override a lot of props just to make it look like NOT a button. There is a way to use a tooltip without a button.
- You do not organize imports, there is a menu item for that in VSCode (right click, source action, organize imports)
- Using keys where there is no need for keys (e.g. not in a loop)
- Not using keys where there IS a need for keys (e.g. in a loop, this was the reason why after I edited some of your code, linter has actually started complaining about missing keys)
- Using `"white"`, when theme default font color is already white. You ideally should not be using `"white"`

How you do it with white:
```
color={bleh ? 'red' : 'white'}
```

How you should be doing it:
```
color={bleh && 'red'}
```

stylemistake committed:
Update Reagents.js

stylemistake committed:
No resizable

stylemistake committed:
Update ReagentLookup.js

stylemistake committed:
Update Reagents.js

Thalpy committed:
Changes UI code as suggested

Thalpy committed:
Few more fixes and tweaks from request

Thalpy committed:
Flexing boxes

Thalpy committed:
No width

Thalpy committed:
No width2

Thalpy committed:
Width to Flex

Thalpy committed:
Fixes width

Thalpy committed:
For clarity

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into reaction-lookup-attempt

Thalpy committed:
Merge remote-tracking branch 'upstream/master'

Thalpy committed:
Merge branch 'master' into reaction-lookup-attempt

Thalpy committed:
Merge remote-tracking branch 'upstream/master' into reaction-lookup-attempt

drongood12 commented:
The window with the reagents opens for a very long time due to the fact that all reagents are being displayed, you probably should have also added pagination, and the search by partial match is also missing. I would also like to see the close button in the Recipe lookup and Reagent lookup menu, because sometimes they interfere with the selection of the desired reagent or recipe

Thalpy commented:
> 
> 
> The window with the reagents opens for a very long time due to the fact that all reagents are being displayed, you probably should have also added pagination, and the search by partial match is also missing. I would also like to see the close button in the Recipe lookup and Reagent lookup menu, because sometimes they interfere with the selection of the desired reagent or recipe
> ![image](https://user-images.githubusercontent.com/73915636/108612153-d9e8e500-73f6-11eb-94dd-d01e4002aae2.png)

I'll see if I can figure out pagination, and add a close button while I'm at it. I'm not sure what you mean by partial match though.

drongood12 commented:
> > The window with the reagents opens for a very long time due to the fact that all reagents are being displayed, you probably should have also added pagination, and the search by partial match is also missing. I would also like to see the close button in the Recipe lookup and Reagent lookup menu, because sometimes they interfere with the selection of the desired reagent or recipe
> > ![image](https://user-images.githubusercontent.com/73915636/108612153-d9e8e500-73f6-11eb-94dd-d01e4002aae2.png)
> 
> I'll see if I can figure out pagination, and add a close button while I'm at it. I'm not sure what you mean by partial match though. The UI was less laggy when I was using it locally unforunately.

By partial coincidence, I meant that you could enter, for example, space and I was given all the chemicals that contain this text. Also why does a chemical dispenser need bartender drinks? Why not make a separation for this, the chemist only has reagents, the barmern only has drinks

Thalpy commented:
> 
> 
> > > The window with the reagents opens for a very long time due to the fact that all reagents are being displayed, you probably should have also added pagination, and the search by partial match is also missing. I would also like to see the close button in the Recipe lookup and Reagent lookup menu, because sometimes they interfere with the selection of the desired reagent or recipe
> > > ![image](https://user-images.githubusercontent.com/73915636/108612153-d9e8e500-73f6-11eb-94dd-d01e4002aae2.png)
> > 
> > 
> > I'll see if I can figure out pagination, and add a close button while I'm at it. I'm not sure what you mean by partial match though. The UI was less laggy when I was using it locally unforunately.
> 
> By partial match, I mean that you can enter, for example, space and all the chemicals in the list containing that text will be displayed. Also why does a chemical dispenser need a bartender's drinks? Why not make a separation for this, the chemist has only reagents, the bartender has only drinks
> 
> P.s
> 
> I may be mistaken in writing some sentences, because English is not my native language. Sorry(

Don't worry! Your english is understandable and I get what you mean since you clarified. To do the search as a list thing might take a bit more time, and I want to make sure some of the more important parts of chemistry are added before I attempt that, but I did add the pagination and closing request you put in. I figured I might as well have all recipies in one place for ease of use and incase bartenders/chemists felt like broadening their horizons since it's meant to be a lookup thing.
